### ISSUE TITLE
The following token types are not defined but have tokens: $info['types']['menu-link']

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1901]
I saw this in my status report and that's what made me decide to scrap my local dev and build a fresh one. The goal was to upgrade 1.3.5 to 1.3.7 and then 1.4.2 and see if this is reproducible, but I've hit #1900. Maybe it means something to someone that knows how to fix it. If I sort the fatal, I'll see if this happens with the fresh install + upgrade to 1.4.2 and if not, I will close this issue here.

**Comment by BWPanda** [comment]
This just happened to me. Turns out it's simply a result of disabling the Menu module (which isn't a required module, so there are use-cases for doing so).

**Comment by alanmels** [comment]
Hitting the same error, but with slightly different message:

> The following token types are not defined but have tokens: $info['types']['term']

Re-enabling the taxonomy module makes it go away, however this website does not need the taxonomy.

I'm using this dirty, but quick hack for now to make the notice go away - replaced the following block in `./includes/token.inc`:

```php
  // Check tokens for problems.
  foreach ($token_info['tokens'] as $type => $tokens) {
    if (!is_array($tokens)) {
      $token_problems['not-array']['problems'][] = "\$info['tokens']['$type']";
      continue;
    }
    else {
      foreach (array_keys($tokens) as $token) {
        if (!is_array($tokens[$token])) {
          $token_problems['not-array']['problems'][] = "\$info['tokens']['$type']['$token']";
          continue;
        }
        elseif (!isset($tokens[$token]['name']) || !isset($tokens[$token]['description'])) {
          $token_problems['missing-info']['problems'][] = "\$info['tokens']['$type']['$token']";
        }
        elseif (is_array($tokens[$token]['name']) || is_array($tokens[$token]['description'])) {
          $token_problems['duplicate']['problems'][] = "\$info['tokens']['$type']['$token']";
        }
      }
    }
    if (!isset($token_info['types'][$type])) {
      $token_problems['tokens-no-type']['problems'][] = "\$info['types']['$type']";
    }
  }
```

with:

```php
  // Check tokens for problems.
  foreach ($token_info['tokens'] as $type => $tokens) {
    if (!is_array($tokens)) {
      $token_problems['not-array']['problems'][] = "\$info['tokens']['$type']";
      continue;
    }
    else {
      foreach (array_keys($tokens) as $token) {
        if (!is_array($tokens[$token])) {
          $token_problems['not-array']['problems'][] = "\$info['tokens']['$type']['$token']";
          continue;
        }
        elseif (!isset($tokens[$token]['name']) || !isset($tokens[$token]['description'])) {
          $token_problems['missing-info']['problems'][] = "\$info['tokens']['$type']['$token']";
        }
        elseif (is_array($tokens[$token]['name']) || is_array($tokens[$token]['description'])) {
          $token_problems['duplicate']['problems'][] = "\$info['tokens']['$type']['$token']";
        }
      }
    }
    if (!isset($token_info['types'][$type])) {
      if ($type == 'term' && !module_exists('taxonomy')) {
        continue;
      }
      $token_problems['tokens-no-type']['problems'][] = "\$info['types']['$type']";
    }
  }
```

but it must be properly fixed in core.

**Comment by argiepiano** [comment]
I'm not able to reproduce. This is what I'm doing:

1. Clean install
2. Delete the Tags field from Post
3. Disable and uninstall Taxonomy
4. Clear caches
... Nothing

Am I missing a step, @alanmels? Otherwise, I guess this must be the result of a contrib module defining tokens of type `'term'` or `'vocabulary'`, and therefore core will complain if taxonomy is disabled, since that's were those token types are defined.

**Comment by alanmels** [comment]
On a clean website disable Taxonomy and enable Tolen module and look at the Status page, and the notice should be there.

**Comment by argiepiano** [comment]
> On a clean website disable Taxonomy and enable Tolen module and look at the Status page, and the notice should be there.

Yes, that's what I'm doing. I don't see a problem in the Status page.

**Comment by argiepiano** [comment]
OK, I was able to reproduce ONLY if:
1. Enable Entity Plus
2. Enable Entity Tokens
3. Disable Taxonomy

So, the problem is not in core. The problem is that Entity Tokens defines a few custom term tokens. I'll open an issue there.

**Comment by alanmels** [comment]
Pardon me, I also have Entity Plus.

**Comment by argiepiano** [comment]
Posted an issue in Entity Tokens:

https://github.com/backdrop-contrib/entity_token/issues/43

**Comment by argiepiano** [comment]
@alanmels, this is now fixed in Entity Tokens, and just did a new release of the module. Please test.

**Comment by klonos** [comment]
In the meantime, I have been working on cross-ports for commits in the 7.x Token module (as part of the #5012 meta issue). This one specifically will affect how the status report messages look (severity etc.): https://github.com/backdrop/backdrop/pull/4489

Example of a status warning before:
![image](https://github.com/backdrop/backdrop/assets/2423362/43438601-f188-453c-8695-c18950511dd2)

After this change it becomes an "info" message:
![image](https://github.com/backdrop/backdrop/assets/2423362/f9ae874a-bd4b-4924-b4f1-983cbdaff336)
